XyzWalker/Assets/Src/Scripts/CodeExecutor.cs

```diff
@@ -0,0 +1,142 @@
+// to Unity's main thread when needed.
+public class CodeExecutor
+    : IVisualizer<Event.SingleStep>,
+      IVisualizer<Event.Assignment> {
+  private class ConsoleWriter : TextWriter {
+    private readonly GameManager _gameManager;
+    private readonly StringBuilder _buffer = new StringBuilder();
+
+    public override System.Text.Encoding Encoding => Encoding.Default;
+
+    public ConsoleWriter(GameManager gameManager) {
+      _gameManager = gameManager;
+    }
+
+    public override void Write(char c) {
+      if (c == '\n') {
+        if (_buffer.Length > 0) {
+          _gameManager.QueueOutputTextInfo(_buffer.ToString());
+          _buffer.Clear();
+        }
+      } else {
+        _buffer.Append(c);
+      }
+    }
+  }
+
+  private const string _defaultModuleName = "Program";
+  private const int _minSleepInMilliSeconds = 30;
+  private const float _singleStepWaitInSeconds = .5f;
+
+  private readonly GameManager _gameManager;
+  private readonly ConsoleWriter _consoleWriter;
+  private readonly object _threadLock = new object();
+
+  // If the executor thread is running.
+  public bool IsRunning => !(_thread is null);
+
+  // A switch for the caller to stop the code execution when the thread is running. It's safe for
+  // another thread to flip this bool flag directly.
+  public bool Stopping = false;
+
+  private Thread _thread;
+  private string _source;
+
+  public CodeExecutor(GameManager gameManager) {
+    _gameManager = gameManager;
+    _consoleWriter = new ConsoleWriter(gameManager);
+  }
+
+  // Runs a SeedLang script. Returns false if the executor is already running.
+  public bool Run(string source) {
+    bool ret = true;
+    lock (_threadLock) {
+      if (_thread is null) {
+        _source = source;
+        Stopping = false;
+        _thread = new Thread(ThreadEntry);
+        _thread.Start();
+      } else {
+        ret = false;
+      }
+    }
+    return ret;
+  }
+
+  public void On(Event.SingleStep e, IVM vm) {
+    // Checks the stopping flag in the SingleStep callback.
+    if (Stopping) {
+      vm.Stop();
+      Stopping = false;
+    } else {
+      // Highlights the current line.
+      _gameManager.QueueHighlightCodeLineAndWait(e.Range.Start.Line, _singleStepWaitInSeconds);
+      WaitForActionQueueComplete();
+    }
+  }
+
+  public void On(Event.Assignment e, IVM vm) {
+    // Monitors the assignment events for the variables "x", "y", "z".
+    if (e.Name == "x") {
+      _gameManager.QueueVisualizeX((float)e.Value.AsNumber());
+    } else if (e.Name == "y") {
+      _gameManager.QueueVisualizeY((float)e.Value.AsNumber());
+    } else if (e.Name == "z") {
+      _gameManager.QueueVisualizeZ((float)e.Value.AsNumber());
+    }
+    WaitForActionQueueComplete();
+  }
+
+  // This method is used to synchronize the executor thread and the main UI thread. For example, it
+  // will be confusing if the main UI thread is still playing the swapping animation while the
```

Comment:
Comments should be changed, as there isn't `swapping animation` in this example?

Reply:
Done.